Arm/Memory/SeparateAutomation.lean

```diff
@@ -697,99 +704,130 @@ partial def SimpMemM.simplifyExpr (e : Expr) (hyps : Array Hypothesis) : SimpMem
+partial def SimpMemM.simplifyGoal (g : MVarId) (hyps : Array Hypothesis) : SimpMemM Bool := do
   SimpMemM.withMainContext do
     trace[simp_mem.info] "{processingEmoji} Matching on ⊢ {← g.getType}"
     let gt ← g.getType
-    if let .some e := MemLegalProp.ofExpr? gt then do
+    if let .some e := MemLegalProp.ofExpr? gt then
       withTraceNode m!"Matched on ⊢ {e}. Proving..." do
-      if let .some proof ← proveWithOmega? e hyps then do
-        (← getMainGoal).assign proof.h
-    if let .some e := MemSubsetProp.ofExpr? gt then do
+        if let .some proof ← proveWithOmega? e hyps then
+          (← getMainGoal).assign proof.h
+      return true
+
+    if let .some e := MemSubsetProp.ofExpr? gt then
       withTraceNode m!"Matched on ⊢ {e}. Proving..." do
-      if let .some proof ←  proveWithOmega? e hyps then do
-        (← getMainGoal).assign proof.h
-    if let .some e := MemSeparateProp.ofExpr? gt then do
+        if let .some proof ←  proveWithOmega? e hyps then
+          (← getMainGoal).assign proof.h
+      return true
+    if let .some e := MemSeparateProp.ofExpr? gt then
       withTraceNode m!"Matched on ⊢ {e}. Proving..." do
-      if let .some proof ←  proveWithOmega? e hyps then do
-        (← getMainGoal).assign proof.h
-    withTraceNode m!"Simplifying goal." do
-      SimpMemM.simplifyExpr (← whnf gt) hyps
-
+        if let .some proof ←  proveWithOmega? e hyps then
+          (← getMainGoal).assign proof.h
+      return true
+    else
+      let changed? ← withTraceNode m!"Simplifying goal." do
+          SimpMemM.simplifyExpr (← whnf gt) hyps
+      return changed?
 end
 
 /--
 The core simplification loop.
 We look for appropriate hypotheses, and simplify (often closing) the main goal using them.
 -/
-def SimpMemM.simplifyLoop : SimpMemM Unit := do
-    (← getMainGoal).withContext do
+partial def SimpMemM.simplifyLoop : SimpMemM Unit := do
+  (← getMainGoal).withContext do
+    let mut madeProgress? := false
+    while true do
       let hyps := (← getLocalHyps)
       let foundHyps ← withTraceNode m!"Searching for Hypotheses" do
         let mut foundHyps : Array Hypothesis := #[]
         for h in hyps do
           foundHyps ← hypothesisOfExpr h foundHyps
         pure foundHyps
+
       withTraceNode m!"Summary: Found {foundHyps.size} hypotheses" do
         for (i, h) in foundHyps.toList.enum do
           trace[simp_mem.info] m!"{i+1}) {h}"
 
-      withTraceNode m!"Performing Rewrite At Main Goal" do
+      if (← getUnsolvedGoals).isEmpty then
+        trace[simp_mem.info] "{checkEmoji} All goals solved."
+        break
+
+      let changed? ← withTraceNode m!"Performing Rewrite At Main Goal" do
         SimpMemM.simplifyGoal (← getMainGoal) foundHyps
+      madeProgress? := madeProgress? || changed?
 
+      if changed?
+      then continue
+        else if !madeProgress? && (← getConfig).failIfUnchanged then
+          throwError "{crossEmoji} simp_mem failed to make progress."
+        break
```

Comment:
I find the indentation confusing. Mind lining the if/then/elses up for me? Thx!

Reply:
I cleaned up the control flow instead :) Hope the new version reads better.